pkg/ingester/owned_streams.go

```diff
@@ -57,3 +76,79 @@ func (s *ownedStreamService) decOwnedStreamCount() {
 	}
 	s.ownedStreamCount--
 }
+
+func (s *ownedStreamService) starting(ctx context.Context) error {
+	return nil
+}
+
+func (s *ownedStreamService) running(ctx context.Context) error {
+	ticker := time.NewTicker(s.ringPollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			ringChanged, err := s.checkRingForChanges()
+			if err != nil {
+				level.Error(s.logger).Log("msg", "failed to check ring for changes", "error", err)
+				continue
+			}
+			if ringChanged {
+				s.updateOwnedStreams(ctx)
+			}
+		}
+	}
+}
+
+func (s *ownedStreamService) stopping(_ error) error {
+	return nil
+}
+
+func (s *ownedStreamService) checkRingForChanges() (bool, error) {
+	rs, err := s.ingestersRing.GetAllHealthy(ownedStreamRingOp)
+	if err != nil {
+		return false, err
+	}
+
+	ringChanged := ring.HasReplicationSetChangedWithoutStateOrAddr(s.previousRing, rs)
+	s.previousRing = rs
+	return ringChanged, nil
+}
+
+func (s *ownedStreamService) getTokenRangesForTenant() (ring.TokenRanges, error) {
+	ranges, err := s.ingestersRing.GetTokenRangesForInstance(s.instanceID)
+	if err != nil {
+		if errors.Is(err, ring.ErrInstanceNotFound) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return ranges, nil
+}
+
+func (s *ownedStreamService) ownerKeyAndValue() (string, string) {
+	return "ingester", s.instanceID
+}
+
+func (s *ownedStreamService) updateOwnedStreams(_ context.Context) {
+	ranges, err := s.getTokenRangesForTenant()
+	if err != nil {
+		k, v := s.ownerKeyAndValue()
+		level.Error(s.logger).Log("msg", "failed to get token ranges for tenant", "error", err, k, v)
+		return
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	newOwnedStreamCount := len(ranges)
```

Comment:
ranges have to be used to check whether the stream is owned by the current ingester or not. it's not a count of the streams.